### PR TITLE
Mark command log record duplicates #4213

### PR DESCRIPTION
**Pull request by gerhardol**
A crude workaround, it seems like Exited is running twice (similar detection in the main routine only see one command).

See issue for further discussion.

Changes proposed in this pull request:
 - Mark "probable" duplicates in the log 
 
Screenshots before and after (if PR changes UI):
- 
image

How did I test this code:
 - I do not know how to trigger the problem with duplicate logs, but it appears in all logs I check
- View the log

Has been tested on (remove any that don't apply):
 - Windows 10

### CODE REVIEW

File: GitCommands/Git/GitCommandHelpers.cs [hunk 1]
```diff
@@ -153,11 +155,21 @@ internal static Process StartProcess(string fileName, string arguments, string w
...
+                //GitHub #4213 Commands are duplicated in GE Gitcommand log
+                string duplicate = "";
+                if (_lastLogArguments == arguments && (_lastLogTimestamp - executionStartTimestamp).TotalMilliseconds == 0)
+                {
+                    duplicate = " #Duplicate?";
+                }
+                _lastLogArguments = arguments;
+                _lastLogTimestamp = executionStartTimestamp;
+
+                AppSettings.GitLog.Log(quotedCmd + " " + arguments + duplicate, executionStartTimestamp, executionEndTimestamp);
```
Comment (RussKie) [inline GitCommands/Git/GitCommandHelpers.cs]: can't we simply unwire the event once it fired?

```
        internal static Process StartProcess(string fileName, string arguments, string workingDirectory, Encoding outputEncoding)
        {
            // ...
            startProcess.EnableRaisingEvents = true;
            EventHandler processExited = null;
            processExited = (sender, args) =>
            {
                startProcess.Exited -= processExited;

                string quotedCmd = fileName;
                if (quotedCmd.IndexOf(' ') != -1)
                    quotedCmd = quotedCmd.Quote();
                var executionEndTimestamp = DateTime.Now;
                AppSettings.GitLog.Log(quotedCmd + " " + arguments, executionStartTimestamp, executionEndTimestamp);
            };
            startProcess.Exited += processExited;

            return startProcess;
        }
```

Here's the result:
image

Comment (gerhardol) [reply GitCommands/Git/GitCommandHelpers.cs]: > can't we simply unwire the event once it fired?

Seem like a better solution. I will need to read on the details before I "approve". 
@RussKie  as you did the solution, I can "approve" a PR from you instead.

**Comment by gerhardol** [comment]
More on the problem:
Occasionally, (sender as Process).ExitTime throws an exception even if HasExited is true (WaitForExit also fails). This is not according to the documentation, but does not seem to be the problem.

With the workaround, the problem is much less frequent. However, I saw it once when randomly clicking around. I suggest the duplicate detection is kept.
I still see the problem though:

image

**Comment by RussKie** [comment]
> Occasionally, (sender as Process).ExitTime throws an exception even if
HasExited is true (WaitForExit also fails).

where/under what circumstance have you observed this?


> I still see the problem though

Is that with your change or mine?

**Comment by gerhardol** [comment]
> where/under what circumstance have you observed this?

Testing, adding calls to HasExited and ExitTime

> Is that with your change or mine?

Your change and my detection. It could also have been that the command was dine twice, no check for duplicate commands.

 I will check the logs randomly throughout the day. If no further occurrences, maybe a comment in the code is sufficient. If larger number, maybe leave it as is.

**Comment by RussKie** [comment]
Here's my version - https://github.com/gitextensions/gitextensions/pull/4221

**Comment by gerhardol** [comment]
I can see duplicates detected still, maybe 1 out of 100 seem to be duplicated - or git rev-parse is really called many times in a row. (Need to add some sequence number too to be sure).
image

**Comment by RussKie** [comment]
I would lean towards the fact that we're constantly running `git rev-parse`. In your case these are called by `GitModule.GitCommonDirectory`, which is likely called by `IndexWatcher.SetFileSystemWatcher`...

**Comment by gerhardol** [comment]
> I would lean towards the fact that we're constantly running git rev-parse.

Maybe. I will continue testing (with a counter) tomorrow

**Comment by RussKie** [comment]
Np
I'm not worried if it doesn't make into 2.51 release.
The release is the priority atm

**Comment by gerhardol** [comment]
Not too much in Git today, but the sequence number is updated so this does not seem to be duplicate logging. The rev-parse in such short timeframe should be investigated.

However, the log shows that some sequence numbers are lost, normally the number one or two before the "duplicate".

The (other) PR should be merged. The duplicate rev-parse and the lost logged commands are separate issues.

image

**Comment by gerhardol** [comment]
stash list handling needs to be less aggressive...

image

**Comment by RussKie** [comment]
> However, the log shows that some sequence numbers are lost, normally the number one or two before the "duplicate". (455 in the example below).

I always felt GE was doing something covertly :)

> The (other) PR should be merged

Do you mean #4221?

> stash list handling needs to be less aggressive...

Do you want to raise an issue for this so we can track it?

**Comment by gerhardol** [comment]
Yes merge #4221 
Enhancements come later

I will add one issue about the "lost" logs. This will possibly add a counter to the log (similar to above).

Separate issue about stash list.

**Comment by RussKie** [comment]
Extracted from https://github.com/gitextensions/gitextensions/pull/4712#issuecomment-377520929

**Comment by RussKie** [comment]
Here's what I get opening GE repo on my work laptop:

image

**Comment by RussKie** [comment]
Here's what I get opening GE repo on my home workstation (W10 x64, ESET IS):

image

**Comment by gerhardol** [comment]
It would have been interesting to see the times for _git-stash list_ too. That takes considerable time also un my home PC.

**Pull request closed without merging**